### DISCUSSION TITLE
Avoir un pdf stylisé

### DISCUSSION THREAD

**Discussion by rachelle22** [dompdf/dompdf#3135] [Q&A]
Bonjour, 

je voudrais savoir comment rajouté un fichier css avec dompdf pour avoir un pdf stylisé comment faire s'il vous plait?

**Comment by bsweeney** [comment]
You add CSS to the document the same way you would with any web page.

Inline:
```
<html>
<head>
  <style>
    /* your styles here */
  </style>
</head>
<body>
  <!-- you content here -->
</body>
</html>
```

External:
```
<html>
<head>
  <link href="external.css" rel="stylesheet">
</head>
<body>
  <!-- your content here -->
</body>
</html>
```

**Reply by rachelle22** [reply]
Ca ne marche pas

**Reply by bsweeney** [reply]
We would need to see an example of what you've tried before we can provide any more guidance.

**Comment by crazydef** [comment]
I'm having trouble with this too.

in PHP, I'm just loading a simple HTML file:
```
$dompdf = new Dompdf\Dompdf();
$dompdf->setPaper( 'A4' );
$dompdf->setBasePath( dirname( $filename ));
$dompdf->loadHtml( file_get_contents( $filename ));
$dompdf->render();

return $dompdf->output();
```

Where $filename is created using `dirname( __FILE ) . '/relative/path/to/html'`.

In the HTML itself, I'm pulling in the CSS with the following:

`<link rel="stylesheet" href="styles.css">`

The CSS file is in the same folder as the HTML file referenced in PHP.

The resulting PDF I get has no styling whatsoever.  (And all images are missing too even though they're currently referenced through full URLs such as:  `http://osocorporation.com/api/osologo.png`)

**Reply by bsweeney** [reply]
You are setting the base path so I would expect that to help Dompdf find the CSS, which is presumably in the same directory as the HTML? You could try using `loadHtmlFile` instead and take out the call to `setBasePath`, though this should have the same effect.

As for the image, harder to say. Full URLs usually work fine unless Dompdf does not have ready/write access to the temp directory or if the URL is redirecting from HTTP to HTTPS since redirects are not followed by default in the latest release.

You can try some options to enable additional messaging. I've summarized steps for this in [Debugging Dompdf](https://github.com/dompdf/dompdf/wiki/Debugging-Dompdf) wiki page.

**Reply by crazydef** [reply]
Thanks for the debugging hints.

The images weren't working because isRemoteEnabled in the default options is set to false.  That's fine though, I can just access them through the file system.

However, this is where things get interesting.  :)

It seems just setting the base path isn't enough.  I also needed to set chroot in Options and pass those options to the Domdpf constructor.

If I set the base path, but don't set chroot, I get errors for both the CSS and the images.
If I set chroot, but don't set the base path, I still get errors for the images.
If I set both everything works - with the caveat that image paths cannot reference directories above chroot.  I.e., I can't specify `../some/image.png`, for example.  So chroot really must be the root directory common to all files you plan on accessing.

But everything seems to be working for me now, thanks.  :)

**Reply by bsweeney** [reply]
smh I should have caught both of those scenarios.

It is true that the isRemoteEnabled option must be set to true to access resources over HTTP ([validator source](https://github.com/dompdf/dompdf/blob/v2.0.3/src/Options.php#L1147-L1158)). It's not a "bad" option to enable you just have to keep in mind your data source and potential for malicious use. Or, if you wanted, you could write a custom validator so that you could implement domain whitelisting.

It is also true that a local resource must be under a valid path specified in the chroot option ([validator source](https://github.com/dompdf/dompdf/blob/v2.0.3/src/Options.php#L1108-L1135)). Note that chroot is an array so you can specify multiple paths if your resource are stored in multiple locations.

**Reply by crazydef** [reply]
I think you might want to look into the relationship between chroot and the base path though.  There's definitely some weirdness going on there.  (Personally, I'd be inclined to ditch the base path all together and just use chroot.  The base path seems more of a red herring than a real option.)

**Reply by bsweeney** [reply]
>  The base path seems more of a red herring than a real option

I don't quite follow, base path and chroot _are_ related in that the base path, when a local path, must be allowed via the chroot option. But otherwise they serve differing purposes. The current logic is fine so long as you have a good understanding of how it works, but there are definitely scenarios that some users find confusing.

Ultimately I think the way we'll address all this is to allow users more control over how paths are parsed by making the resource retrieval process more extensible rather than attempt to accommodate additional scenarios in the code.

**Reply by crazydef** [reply]
As you point out, chroot can be an array of paths.  Which I understood to mean an array of paths that will be searched when trying to locate files.  Which is great.  I have complete control over the paths that are searched and the order in which that search happens.  So what is the base path?

In my tests mentioned above, I got different results depending on whether I set both or one of chroot and the base path.  Crucially, it's important to note, I was setting a single chroot path to the same value as the base path:

```
$options = new Dompdf\Options();
$options->set([ 'chroot' => $dir ]);

$dompdf = new Dompdf\Dompdf( $options );
$dompdf->setBasePath( $dir );
```

> If I set the base path, but don't set chroot, I get errors for both the CSS and the images.
> If I set chroot, but don't set the base path, I still get errors for the images.
> If I set both everything works

So there's definitely some inconsistencies with regards to how paths are handled.  And to that end, in my mind, the base path seems totally redundant.

**Reply by bsweeney** [reply]
chroot is not an array of paths that will be searched, it is an array of paths that are permitted to be used when referencing a resource. chroot in no way impacts the location where Dompdf looks for a resource. That is, however, exactly what base path does.

Let's try this as a conversation...
me: I'd like the file at `images/dompdf.png`
base path: I think you meant `/inet/www/res/images/dompdf.png`
chroot: that path is acceptable

**Reply by crazydef** [reply]
ok.

Then may I suggest you rename `chroot` to something more obvious?

**Reply by bsweeney** [reply]
That has been suggested and we will make a change at some point.
